File: recyclerview-paginated/src/com/hendraanggrian/recyclerview/widget/PaginationAdapterWrapper.java
package com.hendraanggrian.recyclerview.widget;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

final class PaginationAdapterWrapper extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TYPE_PLACEHOLDER = Integer.MAX_VALUE - 50; // magic
    private static final int TYPE_ERROR = Integer.MAX_VALUE - 100; // magic

    private final RecyclerView.Adapter originalAdapter;
    private final PaginatedRecyclerView.BaseAdapter placeholderAdapter;
    private final PaginatedRecyclerView.BaseAdapter errorAdapter;

    private boolean isPlaceholder = true;
    private boolean isError = false;

    PaginationAdapterWrapper(
        RecyclerView.Adapter originalAdapter,
        PaginatedRecyclerView.BaseAdapter placeholderAdapter,
        PaginatedRecyclerView.BaseAdapter errorAdapter
    ) {
        this.originalAdapter = originalAdapter;
        this.placeholderAdapter = placeholderAdapter != null
            ? placeholderAdapter
            : PaginatedRecyclerView.PlaceholderAdapter.DEFAULT;
        this.errorAdapter = errorAdapter != null
            ? errorAdapter
            : PaginatedRecyclerView.ErrorAdapter.DEFAULT;
    }

    RecyclerView.Adapter getOriginalAdapter() {
        return originalAdapter;
    }

    void updateState(boolean isPlaceholder, boolean isError) {
        this.isPlaceholder = isPlaceholder;
        this.isError = isError;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case TYPE_PLACEHOLDER:
                return placeholderAdapter.onCreateViewHolder(parent, viewType);
            case TYPE_ERROR:
                return errorAdapter.onCreateViewHolder(parent, viewType);
            default:
                return originalAdapter.onCreateViewHolder(parent, viewType);
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (isErrorRow(position)) {
            errorAdapter.onBindViewHolder(holder, position);
        } else if (isPlaceholderRow(position)) {
            placeholderAdapter.onBindViewHolder(holder, position);
        } else {
            originalAdapter.onBindViewHolder(holder, position);
        }
    }

    @Override
    public int getItemCount() {
        return isPlaceholder || isError
            ? originalAdapter.getItemCount() + 1
            : originalAdapter.getItemCount();
    }

    @Override
    public int getItemViewType(int position) {
        if (isErrorRow(position)) {
            return TYPE_ERROR;
        } else if (isPlaceholderRow(position)) {
            return TYPE_PLACEHOLDER;
        } else {
            return originalAdapter.getItemViewType(position);
        }
    }

    @Override
    public long getItemId(int position) {
        return isPlaceholderRow(position) || isErrorRow(position)
            ? RecyclerView.NO_ID
            : originalAdapter.getItemId(position);
    }

    @Override
    public void setHasStableIds(boolean hasStableIds) {
        super.setHasStableIds(hasStableIds);
        originalAdapter.setHasStableIds(hasStableIds);
    }

    boolean isPlaceholderRow(int position) {
        return isPlaceholder && position == getPlaceholderRowPosition();
    }

    private int getPlaceholderRowPosition() {
        return isPlaceholder ? getItemCount() - 1 : -1;
    }

    boolean isErrorRow(int position) {
        return isError && position == getErrorRowPosition();
    }

    private int getErrorRowPosition() {
        return isError ? getItemCount() - 1 : -1;
    }
}

File: recyclerview-paginated/src/com/hendraanggrian/widget/PaginationAdapterWrapper.kt
package com.hendraanggrian.widget

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

internal class PaginationAdapterWrapper(
        val actualAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>,
        val loadingAdapter: PaginatedRecyclerView.LoadingAdapter<RecyclerView.ViewHolder>
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val TYPE_LOADING: Int = Integer.MAX_VALUE - 50 // Magic
    }

    var isDisplaying: Boolean = true
        set(displaying) {
            if (field != displaying) {
                field = displaying
                notifyDataSetChanged()
            }
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            if (viewType == TYPE_LOADING) loadingAdapter.onCreateViewHolder(parent, viewType)
            else actualAdapter.onCreateViewHolder(parent, viewType)

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int): Unit =
            if (isLoadingRow(position)) loadingAdapter.onBindViewHolder(holder, position)
            else actualAdapter.onBindViewHolder(holder, position)

    override fun getItemCount(): Int = if (isDisplaying) actualAdapter.itemCount + 1 else actualAdapter.itemCount

    override fun getItemViewType(position: Int): Int = if (isLoadingRow(position)) TYPE_LOADING else actualAdapter.getItemViewType(position)

    override fun getItemId(position: Int): Long = if (isLoadingRow(position)) RecyclerView.NO_ID else actualAdapter.getItemId(position)

    override fun setHasStableIds(hasStableIds: Boolean) {
        super.setHasStableIds(hasStableIds)
        actualAdapter.setHasStableIds(hasStableIds)
    }

    private val loadingRowPosition: Int get() = if (isDisplaying) itemCount - 1 else -1

    internal fun isLoadingRow(position: Int): Boolean = isDisplaying && position == loadingRowPosition
}